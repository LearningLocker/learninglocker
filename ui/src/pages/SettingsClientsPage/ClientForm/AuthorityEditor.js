import React from 'react';
import PropTypes from 'prop-types';
import DebounceInput from 'react-debounce-input';
import { Map } from 'immutable';
import { compose, setPropTypes, shouldUpdate, withHandlers } from 'recompose';

const defaultIfis = new Map({
  mbox: 'mailto:hello@learninglocker.net',
  account: new Map({
    homePage: 'http://learninglocker.net',
    name: 'New Client',
  }),
  openid: 'http://learninglocker.net/openid/example',
  mbox_sha1sum: '53d04bc579b8b4d082ca4a530642845e2c0bfe74',
});

const getCurrentIfi = (authority) => {
  if (authority.has('mbox')) return 'mbox';
  if (authority.has('account')) return 'account';
  if (authority.has('openid')) return 'openid';
  if (authority.has('mbox_sha1sum')) return 'mbox_sha1sum';
};

const enhance = compose(
  setPropTypes({
    authority: PropTypes.instanceOf(Map).isRequired,
    onAuthorityChange: PropTypes.func.isRequired,
  }),
  shouldUpdate((prev, next) => !(
    prev.authority.equals(next.authority)
  )),
  withHandlers({
    handleIfiChange: ({ authority, onAuthorityChange }) => (event) => {
      const ifi = event.target.value;
      const value = defaultIfis.get(ifi);
      onAuthorityChange(new Map({
        objectType: 'Agent',
        name: authority.get('name'),
        [ifi]: value,
      }));
    },
    handleFieldChange: ({ authority, onAuthorityChange }) => (keyPath, event) => {
      const value = event.target.value;
      onAuthorityChange(authority.setIn(keyPath, value));
    },
  })
);

const renderIfiSelector = ({ authority, handleIfiChange }) =>
  <select defaultValue={getCurrentIfi(authority)} onChange={handleIfiChange} >
    <option value="mbox">Email</option>
    <option value="account">Account</option>
    <option value="openid">Open ID</option>
    <option value="mbox_sha1sum">Encrypted Email</option>
  </select>;

const renderFieldEditor = ({ authority, label, keyPath, handleFieldChange }) =>
  <div className="form-group" key={keyPath.join('.')}>
    <label htmlFor={keyPath.join('.')}>{label}</label>
    <DebounceInput
      id={keyPath.join('.')}
      className="form-control"
      debounceTimeout={377}
      value={authority.getIn(keyPath)}
      onChange={event => handleFieldChange(keyPath, event)} />
  </div>;

const renderAccountEditor = ({ authority, handleFieldChange }) =>
  <div>
    {renderFieldEditor({
      authority,
      label: 'Authority Account Home Page',
      keyPath: ['account', 'homePage'],
      handleFieldChange
    })}
    {renderFieldEditor({
      authority,
      label: 'Authority Account Name',
      keyPath: ['account', 'name'],
      handleFieldChange
    })}
  </div>;

const renderIfiEditor = ({ authority, handleFieldChange }) => {
  const ifi = getCurrentIfi(authority);
  const renderField = label => renderFieldEditor({
    authority,
    label,
    keyPath: [ifi],
    handleFieldChange
  });

  switch (ifi) {
    case 'mbox':
      return renderField('Authority Email');
    case 'openid':
      return renderField('Authority Open ID');
    case 'mbox_sha1sum':
      return renderField('Authority Encrypted Email');
    case 'account':
      return renderAccountEditor({ authority, handleFieldChange });
    default:
      return renderField('Authority');
  }
};

const render = ({ authority, handleFieldChange, handleIfiChange }) =>
  <div>
    <div className="form-group">
      {renderIfiSelector({ authority, handleIfiChange })}
    </div>
    <div className="form-group">
      {renderFieldEditor({
        authority,
        label: 'Authority Name',
        keyPath: ['name'],
        handleFieldChange
      })}
    </div>
    <div>
      {renderIfiEditor({ authority, handleFieldChange })}
    </div>
  </div>;

export default enhance(render);
