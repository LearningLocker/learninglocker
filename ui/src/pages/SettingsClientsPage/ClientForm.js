import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Map, List } from 'immutable';
import { compose, withProps, withHandlers, setPropTypes } from 'recompose';
import { map } from 'lodash';
import btoa from 'btoa';
import Checkbox from 'ui/components/Material/Checkbox';
import Dropdown from 'ui/components/Material/Dropdown';
import { withModels, withModel } from 'ui/utils/hocs';
import { API_SCOPES, getXapiScopes } from 'lib/constants/scopes';
import ValidationList from 'ui/components/ValidationList';
import { HelpBlock } from 'ui/pages/SettingsClientsPage/styled';
import { getAppDataSelector } from 'ui/redux/modules/app';
import AuthorityEditor from './AuthorityEditor';

const renderNoStore = () => [{ value: 'No LRSs available', label: 'No LRSs available', disabled: true }];

const renderDefaultStore = () => ({ value: null, label: '~ No store ~' });

const renderStoreItems = stores =>
  stores.valueSeq().map(store => (
    { value: store.get('_id'), label: store.get('title') }
  ));

const renderStoreDropdown = compose(
  withProps({
    schema: 'lrs',
    first: 100
  }),
  withModels
)(({ models, onSelect, selectedId, elemId }) =>
  <Dropdown
    label="LRS (optional)"
    id={elemId}
    value={selectedId || 'default'}
    onChange={onSelect}
    source={
      models.size > 0 ?
        [renderDefaultStore(), ...renderStoreItems(models)] :
        renderNoStore()
    } />
);

const renderAuthorityEditor = (handleAttrChange, model) => {
  const errors = model.getIn(['errors', 'messages', 'authority'], new List());

  return (
    <div id="authorityEditor">
      <AuthorityEditor
        authority={model.get('authority')}
        onAuthorityChange={handleAttrChange.bind(null, 'authority')} />
      {
        !errors.isEmpty() && (
          <span className="help-block">
            <ValidationList errors={errors} />
          </span>
        )
      }
    </div>
  );
};

const renderScopes = (setScopes, onChangeScopes, selectedScopes) =>
  map(selectedScopes, (name, scope) => (
    <Checkbox
      key={scope}
      label={name}
      checked={setScopes.includes(scope)}
      onChange={onChangeScopes.bind(null, scope)} />
  ));

const enhance = compose(
  connect(state => ({
    enableSingleStatementDeletion: getAppDataSelector('ENABLE_SINGLE_STATEMENT_DELETION')(state)
  }), {}),
  withProps(({ model }) => ({
    schema: 'client',
    id: model.get('_id'),
  })),
  withModel,
  withHandlers({
    handleAttrChange: ({ updateModel }) =>
      (attr, value) => {
        updateModel({ path: [attr], value });
      },
  }),
  withHandlers({
    handleTargetValueChange: ({ handleAttrChange }) =>
      (attr, { target: { value } }) => {
        handleAttrChange(attr, value);
      },
    onChangeScopes: ({ model, handleAttrChange }) =>
      (scope, checked) => {
        const scopesSet = model.get('scopes').toSet();
        const newScopes = checked ? scopesSet.add(scope) : scopesSet.delete(scope);
        handleAttrChange('scopes', newScopes.toList());
      },
  }),
  setPropTypes({
    model: PropTypes.instanceOf(Map).isRequired,
    handleAttrChange: PropTypes.func.isRequired,
    handleTargetValueChange: PropTypes.func.isRequired,
    onChangeScopes: PropTypes.func.isRequired,
  })
);

export const render = ({
  model,
  handleAttrChange,
  handleTargetValueChange,
  onChangeScopes,
  enableSingleStatementDeletion
}) => {
  const basicKey = model.getIn(['api', 'basic_key'], '');
  const basicSecret = model.getIn(['api', 'basic_secret'], '');
  const _id = model.get('_id', '');
  const isTrusted = model.get('isTrusted', false);
  const title = model.get('title', '');
  const lrsId = model.get('lrs_id', '');
  const scopes = model.get('scopes', new List());

  return (
    <div className="row">
      <div className="col-md-12" >
        <div className="form-group">
          <label htmlFor="lrsTitleInput">Title</label>
          <input
            className="form-control"
            placeholder="Short name for this client"
            value={title}
            onChange={handleTargetValueChange.bind(null, 'title')} />
        </div>

        <div className="form-group">
          <Checkbox
            label="Enabled?"
            onChange={handleAttrChange.bind(null, 'isTrusted')}
            checked={isTrusted} />
        </div>

        <div className="form-group">
          <label htmlFor="basicKey">Key</label>
          <HelpBlock className="help-block">{basicKey}</HelpBlock>
        </div>
        <div className="form-group">
          <label htmlFor="basicSecret">Secret</label>
          <HelpBlock className="help-block">{basicSecret}</HelpBlock>
        </div>
        <div className="form-group">
          <label htmlFor="basicAuth">
            Use this Basic Auth value in your <code>Authorization</code> header for requests
          </label>
          <HelpBlock className="help-block" style={{ wordWrap: 'break-word' }}>
            Basic {btoa(`${basicKey}:${basicSecret}`)}
          </HelpBlock>
        </div>

        <div className="form-group">
          <label htmlFor="overallScopesCheckboxGroup">Overall Scopes</label>
          {renderScopes(scopes, onChangeScopes, API_SCOPES)}
        </div>

        <div className="page-header">
          <h4>xAPI</h4>
        </div>
        <div className="form-group">
          <div>
            {renderStoreDropdown({
              filter: new Map(),
              elemId: `lrsInput${_id}`,
              selectedId: lrsId,
              onSelect: handleAttrChange.bind(null, 'lrs_id'),
            })}
          </div>
        </div>

        {lrsId &&
          <div>
            <div className="form-group">
              <label htmlFor="xapiScopesCheckboxGroup">Scopes</label>
              {renderScopes(
                scopes,
                onChangeScopes,
                getXapiScopes(
                  enableSingleStatementDeletion
                )
              )}
            </div>
            <div className="form-group">
              <label htmlFor="authorityEditor">Authority</label>
              {renderAuthorityEditor(handleAttrChange, model)}
            </div>
          </div>
        }
      </div>
    </div>
  );
};

export default enhance(render);
