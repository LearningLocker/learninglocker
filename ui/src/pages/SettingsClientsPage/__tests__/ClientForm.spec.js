import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import { Map } from 'immutable';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunkMiddleware from 'redux-thunk';
import ClientForm from '../ClientForm';
import AuthorityEditor from '../AuthorityEditor';

global.__DEVELOPMENT__ = true;
global.__CLIENT__ = false;
const createMockStore = configureMockStore([thunkMiddleware]);
const store = createMockStore({
  app: new Map({ ENABLE_STATEMENT_DELETION: false }),
  models: new Map(),
  metadata: new Map(),
  pagination: new Map()
});

describe('ClientForm', () => {
  test('ClientForm renders correctly', () => {
    const tree = ReactTestRenderer.create(
      <Provider store={store}>
        <ClientForm model={new Map({ _id: 'testid' })} />
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('AuthorityEditor renders correctly with mbox', () => {
    const authority = new Map({
      mbox: 'mailto:hello@learninglocker.net',
      name: 'test name'
    });
    const tree = ReactTestRenderer.create(
      <Provider store={store}>
        <AuthorityEditor authority={authority} onAuthorityChange={() => { }} />
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('AuthorityEditor renders correctly with account', () => {
    const authority = new Map({
      account: new Map({
        homePage: 'http://learninglocker.net',
        name: 'New Client',
      }),
      name: 'test name'
    });
    const tree = ReactTestRenderer.create(
      <Provider store={store}>
        <AuthorityEditor authority={authority} onAuthorityChange={() => { }} />
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('AuthorityEditor renders correctly with openid', () => {
    const authority = new Map({
      openid: 'http://learninglocker.net/openid/example',
      name: 'test name'
    });
    const tree = ReactTestRenderer.create(
      <Provider store={store}>
        <AuthorityEditor authority={authority} onAuthorityChange={() => { }} />
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('AuthorityEditor renders correctly with mboxsha1sum', () => {
    const mboxSha1SumAuthority = new Map({
      mbox_sha1sum: '53d04bc579b8b4d082ca4a530642845e2c0bfe74',
      name: 'test name'
    });
    const mboxSha1SumTree = ReactTestRenderer.create(
      <Provider store={store}>
        <AuthorityEditor authority={mboxSha1SumAuthority} onAuthorityChange={() => { }} />
      </Provider>
    ).toJSON();
    expect(mboxSha1SumTree).toMatchSnapshot();
  });
});
