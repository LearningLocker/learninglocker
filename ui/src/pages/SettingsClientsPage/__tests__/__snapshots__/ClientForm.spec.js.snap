// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientForm AuthorityEditor renders correctly with account 1`] = `
<div>
  <div
    className="form-group"
  >
    <select
      defaultValue="account"
      onChange={[Function]}
    >
      <option
        value="mbox"
      >
        Email
      </option>
      <option
        value="account"
      >
        Account
      </option>
      <option
        value="openid"
      >
        Open ID
      </option>
      <option
        value="mbox_sha1sum"
      >
        Encrypted Email
      </option>
    </select>
  </div>
  <div
    className="form-group"
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="name"
      >
        Authority Name
      </label>
      <input
        className="form-control"
        id="name"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        type="text"
        value="test name"
      />
    </div>
  </div>
  <div>
    <div>
      <div
        className="form-group"
      >
        <label
          htmlFor="account.homePage"
        >
          Authority Account Home Page
        </label>
        <input
          className="form-control"
          id="account.homePage"
          onBlur={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          type="text"
          value="http://learninglocker.net"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="account.name"
        >
          Authority Account Name
        </label>
        <input
          className="form-control"
          id="account.name"
          onBlur={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          type="text"
          value="New Client"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ClientForm AuthorityEditor renders correctly with mbox 1`] = `
<div>
  <div
    className="form-group"
  >
    <select
      defaultValue="mbox"
      onChange={[Function]}
    >
      <option
        value="mbox"
      >
        Email
      </option>
      <option
        value="account"
      >
        Account
      </option>
      <option
        value="openid"
      >
        Open ID
      </option>
      <option
        value="mbox_sha1sum"
      >
        Encrypted Email
      </option>
    </select>
  </div>
  <div
    className="form-group"
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="name"
      >
        Authority Name
      </label>
      <input
        className="form-control"
        id="name"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        type="text"
        value="test name"
      />
    </div>
  </div>
  <div>
    <div
      className="form-group"
    >
      <label
        htmlFor="mbox"
      >
        Authority Email
      </label>
      <input
        className="form-control"
        id="mbox"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        type="text"
        value="mailto:hello@learninglocker.net"
      />
    </div>
  </div>
</div>
`;

exports[`ClientForm AuthorityEditor renders correctly with mboxsha1sum 1`] = `
<div>
  <div
    className="form-group"
  >
    <select
      defaultValue="mbox_sha1sum"
      onChange={[Function]}
    >
      <option
        value="mbox"
      >
        Email
      </option>
      <option
        value="account"
      >
        Account
      </option>
      <option
        value="openid"
      >
        Open ID
      </option>
      <option
        value="mbox_sha1sum"
      >
        Encrypted Email
      </option>
    </select>
  </div>
  <div
    className="form-group"
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="name"
      >
        Authority Name
      </label>
      <input
        className="form-control"
        id="name"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        type="text"
        value="test name"
      />
    </div>
  </div>
  <div>
    <div
      className="form-group"
    >
      <label
        htmlFor="mbox_sha1sum"
      >
        Authority Encrypted Email
      </label>
      <input
        className="form-control"
        id="mbox_sha1sum"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        type="text"
        value="53d04bc579b8b4d082ca4a530642845e2c0bfe74"
      />
    </div>
  </div>
</div>
`;

exports[`ClientForm AuthorityEditor renders correctly with openid 1`] = `
<div>
  <div
    className="form-group"
  >
    <select
      defaultValue="openid"
      onChange={[Function]}
    >
      <option
        value="mbox"
      >
        Email
      </option>
      <option
        value="account"
      >
        Account
      </option>
      <option
        value="openid"
      >
        Open ID
      </option>
      <option
        value="mbox_sha1sum"
      >
        Encrypted Email
      </option>
    </select>
  </div>
  <div
    className="form-group"
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="name"
      >
        Authority Name
      </label>
      <input
        className="form-control"
        id="name"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        type="text"
        value="test name"
      />
    </div>
  </div>
  <div>
    <div
      className="form-group"
    >
      <label
        htmlFor="openid"
      >
        Authority Open ID
      </label>
      <input
        className="form-control"
        id="openid"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        type="text"
        value="http://learninglocker.net/openid/example"
      />
    </div>
  </div>
</div>
`;

exports[`ClientForm ClientForm renders correctly 1`] = `
<div
  className="row"
>
  <div
    className="col-md-12"
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="lrsTitleInput"
      >
        Title
      </label>
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="Short name for this client"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        className="sc-bdVaJa DpnOa"
        data-react-toolbox="checkbox"
        onMouseEnter={undefined}
        onMouseLeave={undefined}
      >
        <input
          checked={false}
          className={undefined}
          disabled={false}
          name={undefined}
          onChange={[Function]}
          onClick={[Function]}
          type="checkbox"
        />
        <div
          className="check"
          data-react-toolbox="check"
          onMouseDown={[Function]}
          style={undefined}
        />
        <span
          className={undefined}
          data-react-toolbox="label"
        >
          Enabled?
        </span>
      </label>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="basicKey"
      >
        Key
      </label>
      <p
        className="sc-bxivhb fZLCcU help-block"
      >
        
      </p>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="basicSecret"
      >
        Secret
      </label>
      <p
        className="sc-bxivhb fZLCcU help-block"
      >
        
      </p>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="basicAuth"
      >
        Use this Basic Auth value in your 
        <code>
          Authorization
        </code>
         header for requests
      </label>
      <p
        className="sc-bxivhb fZLCcU help-block"
        style={
          Object {
            "wordWrap": "break-word",
          }
        }
      >
        Basic 
        Og==
      </p>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="overallScopesCheckboxGroup"
      >
        Overall Scopes
      </label>
      <label
        className="sc-bdVaJa DpnOa"
        data-react-toolbox="checkbox"
        onMouseEnter={undefined}
        onMouseLeave={undefined}
      >
        <input
          checked={false}
          className={undefined}
          disabled={false}
          name={undefined}
          onChange={[Function]}
          onClick={[Function]}
          type="checkbox"
        />
        <div
          className="check"
          data-react-toolbox="check"
          onMouseDown={[Function]}
          style={undefined}
        />
        <span
          className={undefined}
          data-react-toolbox="label"
        >
          API All
        </span>
      </label>
      <label
        className="sc-bdVaJa DpnOa"
        data-react-toolbox="checkbox"
        onMouseEnter={undefined}
        onMouseLeave={undefined}
      >
        <input
          checked={false}
          className={undefined}
          disabled={false}
          name={undefined}
          onChange={[Function]}
          onClick={[Function]}
          type="checkbox"
        />
        <div
          className="check"
          data-react-toolbox="check"
          onMouseDown={[Function]}
          style={undefined}
        />
        <span
          className={undefined}
          data-react-toolbox="label"
        >
          Delete statements (not covered by All scope)
        </span>
      </label>
    </div>
    <div
      className="page-header"
    >
      <h4>
        xAPI
      </h4>
    </div>
    <div
      className="form-group"
    >
      <div>
        <div
          className="sc-bwzfXH jLULoW"
          data-react-toolbox="dropdown"
          onBlur={[Function]}
          onFocus={[Function]}
          tabIndex="-1"
        >
          <div
            className=""
            data-react-toolbox="input"
          >
            <input
              className=""
              defaultValue={undefined}
              disabled={false}
              id="lrsInput"
              maxLength={undefined}
              name={undefined}
              onChange={[Function]}
              onClick={[Function]}
              onKeyPress={undefined}
              readOnly={true}
              required={false}
              role="input"
              tabIndex="0"
              type={null}
              value=""
            />
            <span
              className={undefined}
            />
            <label
              className=""
            >
              LRS (optional)
            </label>
          </div>
          <ul
            className="values"
          >
            <li
              className="undefined"
              onMouseDown={false}
            >
              No LRSs available
            </li>
          </ul>
        </div>
      </div>
    </div>
    
  </div>
</div>
`;
